### PR TITLE
Segmentation

### PR DESCRIPTION
**Pull request by isaac-tfn**
Describe Changes
---

Task Checklist
----
- [ ] Have unittests been added (testing individual functions and their edge cases)
- [ ] Have integration tests been added (if applicable, to test modules of functionality)
- [ ] Updated RELEASE.md to reflect changes in PR
- [ ] Have new dependencies been added?
  - [ ] Have they been added to either `requirements.txt` or `requirements_dev.txt`.
  - [ ] Have they been added to `setup.cfg`

### CODE REVIEW

**Review by wsp-mbuckley — COMMENTED** [review]
I think most of my comments don't need fixing now and could just be added as issues (I can add those if needed). The main things I've noticed are the following:

- Segment default values should be dict / list to avoid needing all the None checks everywhere
- Docstrings should use numpy format
- Some type annotations are missing, mostly return types but also Literal strings where relevant

I haven't looked at the tests but can if needed.

File: src/caf/core/zoning.py [hunk 1]
```diff
@@ -0,0 +1,469 @@
...
+        self.unique_zones = unique_zones
+        self.n_zones = len(self.unique_zones)
+        if isinstance(metadata, PathLike):
+            self.metadata = ZoningSystemMetaData.load_yaml(metadata)
+        else:
+            self.metadata = metadata
+
+        # Validate and assign the optional arguments
+        self.internal_zones = internal_zones
+        self.external_zones = external_zones
```
[the remark this answers is not in the document]
Comment (wsp-mbuckley) [reply src/caf/core/zoning.py:106]: Issue #5

File: src/caf/core/zoning.py [hunk 2]
```diff
@@ -0,0 +1,469 @@
...
+import h5py
+import caf.toolkit as ctk
+
+
+# Local Imports
+
+
+LOG = logging.getLogger(__name__)
+
+# This is temporary, and will be an enviornment variable
```
[the remark this answers is not in the document]
Comment (wsp-mbuckley) [reply src/caf/core/zoning.py:38]: issue #6

File: src/caf/core/data_structures.py [hunk 3]
```diff
@@ -0,0 +1,655 @@
...
+            return False
+        if not self.data.equals(other.data):
+            return False
+        return True
+
+    def __ne__(self, other):
+        """Note equals dunder for DVector"""
+        return not self.__eq__(other)
+
+    def aggregate(self, segs: list[str]):
```
[the remark this answers is not in the document]
Comment (wsp-mbuckley) [reply src/caf/core/data_structures.py:631]: issue #7

**Comment by wsp-mbuckley** [comment]
> > I think most of my comments don't need fixing now and could just be added as issues (I can add those if needed). The main things I've noticed are the following:
> > 
> > * Segment default values should be dict / list to avoid needing all the None checks everywhere
> > * Docstrings should use numpy format
> > * Some type annotations are missing, mostly return types but also Literal strings where relevant
> > 
> > I haven't looked at the tests but can if needed.
> 
> Definitely need to improve documentation and annotations. On the None type checks, @BenTaylor-TfN has told me it's bad practice to default to lists/dicts (mutable types) which is why often they default to None. It's possible I'm misunderstanding this comment or I've misunderstood Ben at some point!

@BenTaylor-TfN is correct, but pydantic has a Field class which can have default factory functions which it will call to instantiate a new list each time the class is instantiated it works similar to having functions default to None then inside the function doing
```python
if variable is None:
    variable = []
```
In the comments by the code I've put an example of how to use this pydantic class. The other option is to use pydantic's validator with `always=True` to set to an empty list if the value is None but this is less neat.

The reason I think they shouldn't be allowed to be None is that there are lots of None checks which clutter up other pieces of code and they could be removed if the list / dict weren't allowed to be None, it wouldn't need any empty list / dict checks either because iterating through the empty collection just wouldn't have any iterations.

Can setup a call to go through this and other comments if you like?

File: src/caf/core/data_structures.py [hunk 4]
```diff
@@ -0,0 +1,655 @@
...
+        self._time_format = self._validate_time_format(time_format)
+
+        # Set defaults if args not set
+        val_col = self._val_col if val_col is None else val_col
+
+        # Try to convert the given data into DVector format
+        if isinstance(import_data, pd.DataFrame) or isinstance(import_data, pd.Series):
+            self._data = self._dataframe_to_dvec(import_data)
+        elif isinstance(import_data, dict):
+            self._data = self._old_to_new_dvec(import_data)
```
[the remark this answers is not in the document]
Comment (isaac-tfn) [reply src/caf/core/data_structures.py:276]: Do you mean as a class method but still called here, or as a class method and the init only accepts dataframes?

**Comment by isaac-tfn** [comment]
> Can setup a call to go through this and other comments if you like?

Sounds good thanks - doesn't look like I have many calls Thursday Friday so slot something in.

File: src/caf/core/data_structures.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (wsp-mbuckley) [reply src/caf/core/data_structures.py:276]: class method and the init only accepts dataframes, I think it will make the init clearer and simpler and I don't think people will need the dict functionality very often anyway. Tbh it's mostly my preference I like having an init with a single defined way of initialising the class then any other conversions from other datatypes can just have there own class method e.g. `from_dict`. This is one of the comments that doesn't necessarily need changing.

File: src/caf/core/segmentation.py [hunk 5]
```diff
@@ -0,0 +1,415 @@
...
+    def ind(self):
+        """
+        Returns a pandas MultiIndex of the segmentation. This is by default just a product
+        of all segments given, taking exclusions into account if any exist between segments.
+        """
+        index = pd.MultiIndex.from_product(self.seg_vals, names=self.names)
+        df = pd.DataFrame(index=index)
+        drop_iterator = self.naming_order.copy()
+        for own_seg in self.segments:
+            for other_seg in drop_iterator:
```
[the remark this answers is not in the document]
Comment (isaac-tfn) [reply src/caf/core/segmentation.py:167]: I think at one point the iterator was updated as it was iterated through, the structure got changes but I didn't update that bit

File: src/caf/core/segmentation.py [hunk 6]
```diff
@@ -0,0 +1,415 @@
...
+    def __copy__(self):
+        """Returns a copy of this class"""
+        return self.copy()
+
+    def __eq__(self, other) -> bool:
+        """Overrides the default implementation"""
+        if not isinstance(other, Segmentation):
+            return False
+
+        if set(self.naming_order) != set(other.naming_order):
```
[the remark this answers is not in the document]
Comment (isaac-tfn) [reply src/caf/core/segmentation.py:318]: Yes don't know why that's a set rather than just the lists

File: src/caf/core/zoning.py [hunk 7]
```diff
@@ -0,0 +1,469 @@
...
+    """
+
+    def __init__(
+        self,
+        name: str,
+        unique_zones: pd.DataFrame,
+        metadata: Union[ZoningSystemMetaData, PathLike],
+        zone_descriptions: Optional[pd.DataFrame] = None,
+        internal_zones: Optional[pd.DataFrame] = None,
+        external_zones: Optional[pd.DataFrame] = None,
```
[the remark this answers is not in the document]
Comment (isaac-tfn) [reply src/caf/core/zoning.py:69]: Think I mentioned but this was really just trying to mimic the existing zoning system. It should be simplified to ZoneingSystem being just a dataframe with metadata, and the metadata provides names of extra columns and possibly descriptions (internal: Zones which are internal to TfN for example)

File: src/caf/core/zoning.py [hunk 8]
```diff
@@ -0,0 +1,469 @@
...
+        """A Dictionary of zones to their names"""
+        return self.unique_zones.join(self.zone_descriptions)
+
+    def __copy__(self):
+        """Returns a copy of this class"""
+        return self.copy()
+
+    def __eq__(self, other) -> bool:
+        """Overrides the default implementation"""
+        # May need to update in future, but assume they are equal if names match
```
[the remark this answers is not in the document]
Comment (isaac-tfn) [reply src/caf/core/zoning.py:121]: Think I copied this from normits, edited and didn't remove the comment yes

File: src/caf/core/zoning.py [hunk 9]
```diff
@@ -0,0 +1,469 @@
...
+        other: ZoningSystem,
+        weighting: str = "spatial",
+        trans_cache: Path = Path(r"I:\Data\Zone Translations\cache"),
+    ) -> pd.DataFrame:
+        """
+        Returns a space generate zone translation between self and other.
+        """
+        # Init
+        names = sorted([self.name, other.name])
+        folder = f"{names[0]}_{names[1]}"
```
[the remark this answers is not in the document]
Comment (isaac-tfn) [reply src/caf/core/zoning.py:159]: This is the way they're generated by space and for now space translations are the only ones which should be loaded automatically

File: src/caf/core/data_structures.py [hunk 10]
```diff
@@ -0,0 +1,655 @@
...
+    def overlap(self, other):
+        """Calls segmentation overlap method to check two DVector's contain
+        overlapping segments"""
+        overlap = self.segmentation.overlap(other.segmentation)
+        if overlap == []:
+            raise NotImplementedError(
+                "There are no common segments between the "
+                "two DVectors so this operation is not "
+                "possible."
+            )
```
[the remark this answers is not in the document]
Comment (isaac-tfn) [reply src/caf/core/data_structures.py:552]: Just a check

File: src/caf/core/zoning.py [hunk 11]
```diff
@@ -0,0 +1,469 @@
...
+            raise ValueError("Mode can only be 'hdf' or 'csv', not " f"{mode}.")
+        int_zones = None
+        ext_zones = None
+        descriptions = None
+        if "internal" in zoning.columns:
+            int_zones = zoning.loc[zoning["internal"], "zone_name"]
+        if "external" in zoning.columns:
+            ext_zones = zoning.loc[zoning["external"], "zone_name"]
+        if "descriptions" in zoning.columns:
+            descriptions = zoning["descriptions"]
```
[the remark this answers is not in the document]
Comment (isaac-tfn) [reply src/caf/core/zoning.py:361]: There is already an issue partly covering new zoning stuff but I think an issue just for an overall refactor would be a good idea

File: src/caf/core/zoning.py [hunk 12]
```diff
@@ -0,0 +1,469 @@
...
+            be contained in unique_zones.
+
+        external_zones:
+            A numpy array of unique zone names that make up the "external"
+            area of this zoning system. Every value in this array must also
+            be contained in unique_zones.
+        """
+        # Init
+        self.name = name
+        self.unique_zones = unique_zones
```
[the remark this answers is not in the document]
Comment (isaac-tfn) [reply src/caf/core/zoning.py:89]: I think this should go with the refactor, could you link it to an issue on that when you make it @wsp-mbuckley ?

File: src/caf/core/data_structures.py [hunk 13]
```diff
@@ -0,0 +1,655 @@
...
+            if (
+                set(import_data.columns) != set(self.zoning_system.unique_zones["zone_name"])
+            ) & (set(import_data.columns) != set(self.zoning_system.unique_zones["zone_id"])):
+                raise ImportError(
+                    "The input dataframe does not contain columns matching the zoning system given"
+                )
+        else:
+            import_data.columns = [self.val_col]
+
+        return import_data
```
[the remark this answers is not in the document]
Comment (isaac-tfn) [reply src/caf/core/data_structures.py:374]: Add as issue - I really should do this now but need to get onto Nhan's stuff properly

**Comment by wsp-mbuckley** [comment]
I've added a few issues which should cover all the comments I've made previously #11, #12, #13, #14, #15

**Comment by isaac-tfn** [comment]
@wsp-mbuckley Still some linting stuff - a lot of pydocstyle, some mypy, pylint is all passing. I'm happy to pick it up after Christmas if you don't have time, should take a day or so just don't have time now.

**Comment by wsp-mbuckley** [comment]
> @wsp-mbuckley Still some linting stuff - a lot of pydocstyle, some mypy, pylint is all passing. I'm happy to pick it up after Christmas if you don't have time, should take a day or so just don't have time now.

@isaac-tfn thanks for that, I've got some viz stuff to do next week but if a get time I'll try a get some of this done too so its ready for merging in the new year
